components/portal/messages/controllers.js

```diff
@@ -51,42 +50,48 @@ define(['angular'], function(angular) {
               } else if (angular.isString(result)) {
                 $scope.messagesError = result;
               }
-              filterMessages();
+              filterMessages(allMessages);
               return allMessages;
             })
             .catch(function(error) {
               $log.warn('Problem getting all messages for messages controller');
               return error;
             });
         };
+        // Promise to get seen message IDs
+        var promiseSeenMessageIds = function() {
+          return messagesService.getSeenMessageIds();
+        };
+
+        // Promise to resolve urls in messages
+        var promiseMessagesByData = function(messages) {
+          return messagesService.getMessagesByData(messages);
+        };
+
+        // Promise to resolve group memberships
+        var promiseMessagesByGroup = function(messages) {
+          return messagesService.getMessagesByGroup(messages);
+        };
         /**
          * Determine whether or not messages need to be filtered
          * by group and data, then execute the relevant promises
+         * @param {Object} allMessages
          */
-        var filterMessages = function() {
+        var filterMessages = function(allMessages) {
           var groupFiltersEnabled =
             !$localStorage.disableGroupFilteringForMessages;
-          if (groupFiltersEnabled) {
-            promiseFilteredMessages = {
-              filteredByGroup:
-                messagesService.getMessagesByGroup(allMessages),
-              filteredByData:
-                messagesService.getMessagesByData(allMessages),
-            };
-          }
-
+          var groupPromise = null;
           if (!groupFiltersEnabled) {
-            promiseFilteredMessages = {
-              filteredByGroup: allMessages,
-              filteredByData:
-                messagesService.getMessagesByData(allMessages),
-            };
+            groupPromise = $q.resolve(allMessages);
+          } else {
+            groupPromise = promiseMessagesByGroup(allMessages);
           }
-            // Call filtered notifications promises, then pass on to
-            // the completion function
-            $q.all(promiseFilteredMessages)
-              .then(filterMessagesSuccess)
-              .catch(filterMessagesFailure);
+
+          $q.all([promiseSeenMessageIds(),
+            promiseMessagesByData(allMessages),
+            groupPromise])
+            .then(filterMessagesSuccess)
```

Comment:
:art: It took me a solid minute to recognize that the `then` is not inside the `$q.all`. Maybe some indentation improvements could be made?